File: zmqserver.jl
using HttpServer
using WebSockets
using ZMQ

function handle_file_request(req, res)
    @show req.resource
    if req.resource == "/static/"
        open("test.html") do file
            return Response(read(file))
        end
    else
        return Response(404)
    end
end

function run_server()
    connections = Set{WebSocket}()

    websocket_handler = WebSocketHandler() do req, client
        @show client
        push!(connections, client)
        try
            while true
                read(client)
            end
        catch e
            if e isa WebSockets.WebSocketClosedError
                if client in connections
                    delete!(connections, client)
                end
            end
        end
    end

    server = Server(HttpHandler(handle_file_request),
                    websocket_handler)
    yield()

    context = Context()
    socket = Socket(context, REP)
    ZMQ.bind(socket, "tcp://*:5555")

    @async run(server, 8888)
    while true
        msg = take!(convert(IOStream, ZMQ.recv(socket)))
        @show msg
        for connection in connections
            write(connection, msg)
        end
        ZMQ.send(socket, "ok")
    end
end

run_server()

File: zmqserver.py
from __future__ import absolute_import, division, print_function

import os

import tornado.web
import tornado.ioloop
import tornado.websocket

import zmq
from zmq.eventloop import ioloop
from zmq.eventloop.zmqstream import ZMQStream

# Install ZMQ ioloop instead of a tornado ioloop
# http://zeromq.github.com/pyzmq/eventloop.html
ioloop.install()

connections = set()


class WebSocketHandler(tornado.websocket.WebSocketHandler):
    def open(self):
        connections.add(self)
        print("opened:", self)

    def on_message(self, message):
        print("got message:", message)

    def on_close(self):
        connections.remove(self)
        print("closed:", self)


def setup_zmq():
    context = zmq.Context()
    socket = context.socket(zmq.REP)
    socket.bind("tcp://*:5555")
    stream = ZMQStream(socket)

    def publish_to_websockets(msg):
        print("msg:", msg)
        for websocket in connections:
            websocket.write_message(msg[0])
        socket.send(b"ok")

    stream.on_recv(publish_to_websockets)
    return stream

def make_app():
    return tornado.web.Application([
        (r"/static/(.*)", tornado.web.StaticFileHandler, {"path": os.path.dirname(__file__), "default_filename": "test.html"}),
        (r"/", WebSocketHandler)
    ])

if __name__ == '__main__':
    stream = setup_zmq()
    app = make_app()
    app.listen(8888)

    import webbrowser
    webbrowser.open("http://localhost:8888/static/")

    tornado.ioloop.IOLoop.current().start()
